core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,325 @@
+    case BTN_THUMBR:
+        return WPE_GAMEPAD_BUTTON_RIGHT_STICK;
+    case BTN_DPAD_UP:
+        return WPE_GAMEPAD_BUTTON_D_PAD_TOP;
+    case BTN_DPAD_DOWN:
+        return WPE_GAMEPAD_BUTTON_D_PAD_BOTTOM;
+    case BTN_DPAD_LEFT:
+        return WPE_GAMEPAD_BUTTON_D_PAD_LEFT;
+    case BTN_DPAD_RIGHT:
+        return WPE_GAMEPAD_BUTTON_D_PAD_RIGHT;
+    case BTN_MODE:
+        return WPE_GAMEPAD_BUTTON_CENTER;
+    default:
+        return -1;
+    }
+}
+
+static void
+on_button_press(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    uint16_t button;
+    int      std_button;
+
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    std_button = toStandardGamepadButton(button);
+    if (std_button != -1)
+        wpe_gamepad_dispatch_button_changed(gamepad->server, (enum wpe_gamepad_button) std_button, true);
+}
+
+static void
+on_button_release(ManetteDevice *device, ManetteEvent *event, Gamepad *gamepad)
+{
+    uint16_t button;
+    if (!manette_event_get_button(event, &button))
+        return;
+
+    wpe_gamepad_dispatch_button_changed(gamepad->server, toStandardGamepadButton(button), false);
+}
+
+static void
+device_disconnect_cb(ManetteDevice *device, Gamepad *gamepad)
+{
+    if (!gamepad)
+        return;
+
+    if (gamepad->button_pressed > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_pressed);
+        gamepad->button_pressed = 0;
+    }
+
+    if (gamepad->button_released > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->button_released);
+        gamepad->button_released = 0;
+    }
+
+    if (gamepad->axis_event > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->axis_event);
+        gamepad->axis_event = 0;
+    }
+
+    if (gamepad->disconnected > 0) {
+        g_signal_handler_disconnect(gamepad->device, gamepad->disconnected);
+        gamepad->disconnected = 0;
+    }
+}
+
+static gpointer
+gamepad_create(struct wpe_gamepad *server, unsigned gamepad_id)
+{
+    Gamepad       *gamepad;
+    ManetteDevice *device;
+
+    device = GUINT_TO_POINTER(gamepad_id);
+    if (!MANETTE_IS_DEVICE(device))
+        return NULL;
+
+    gamepad = g_new0(Gamepad, 1);
+
+    gamepad->server = server;
+    gamepad->device = g_object_ref(device);
+
+    gamepad->button_pressed = g_signal_connect(device, "button-press-event", G_CALLBACK(on_button_press), gamepad);
+    gamepad->button_released = g_signal_connect(device, "button-release-event", G_CALLBACK(on_button_release), gamepad);
+    gamepad->axis_event = g_signal_connect(device, "absolute-axis-event", G_CALLBACK(on_absolute_axis), gamepad);
+    gamepad->disconnected = g_signal_connect(device, "disconnected", G_CALLBACK(device_disconnect_cb), gamepad);
+
+    g_message("gamepad: %s - created", manette_device_get_name(device));
+
+    return gamepad;
+}
+
+static void
+gamepad_destroy(gpointer data)
+{
+    Gamepad *gamepad = data;
+
+    if (!(gamepad && gamepad->device))
```

Comment:
It looks to me that `gamepad_destroy()` is always called with a valid pointer and that there are unneeded checks here. Remember that `g_clear_object()` will only unref if the pointer is non-null so the check on `gamepad->device` is definitely unneeded.

Reply:
done